src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs

```diff
@@ -1431,6 +1496,7 @@ private void UpdateConnectionState()
             {
                 if (TryClose())
                 {
+                    SetConnectionErrorCode(_gracefulCloseReason, Http2ErrorCode.NO_ERROR);
```

Comment:
Is there a race between writing and reading `_gracefulCloseReason`?

Reply:
I don't think so because it is only set if graceful close hasn't already started. After it is set, the connection read loop is canceled, at which point it can be read.